### PR TITLE
feat: add clickhouse and greenplum support

### LINKED ISSUE #459: feat: support Clickhouse and Greenplum DBs
### Describe the feature
Add support for new DBs: [Clickhouse](https://clickhouse.com), [Greenplum](https://greenplum.org)

### Describe alternatives you've considered
-

### Additional context
-

### Who will this benefit?
Any users or above mentioned databases

### Are you interested in contributing this feature?
sure thing!

### PR DESCRIPTION
**Pull request by igorvoltaic**
This is a:
- [ ] bug fix PR with no breaking changes
- [x] new functionality

## Link to Issue 
Closes #459

## Description & motivation
Let's Make dbt-clickhouse users happy ^_^
We are using this in our prod environments and have decided to open source the feature

## Integration Test Screenshot
Screenshot 2024-06-12 at 14 35 56
Screenshot 2024-06-12 at 14 35 42
Screenshot 2024-06-12 at 14 35 11

## Checklist
- [x] I have verified that these changes work locally on the following warehouses (Note: it's okay if you do not have access to all warehouses, this helps us understand what has been covered)
    - [x] Clickhouse
    - [x] Greenplum (Postgres)
    - [ ] Redshift
    - [ ] Snowflake
    - [ ] Databricks
    - [ ] DuckDB
    - [ ] Trino/Starburst
- [x] I have updated the README.md (if applicable)

### CODE REVIEW

File: macros/unpack/get_exposure_values.sql [hunk 1]
```diff
@@ -17,7 +17,7 @@
               wrap_string_with_quotes(node.name),
               wrap_string_with_quotes(node.resource_type),
               wrap_string_with_quotes(node.original_file_path | replace("\\","\\\\")),
-              "cast(" ~ dbt_project_evaluator.is_not_empty_string(node.description) | trim ~ " as boolean)",
+              "cast(" ~ dbt_project_evaluator.is_not_empty_string(node.description) | trim ~ " as " ~ dbt.type_boolean()),
```
Comment (b-per) [inline macros/unpack/get_exposure_values.sql:20]: Wouldn't this need to be:

```suggestion
              "cast(" ~ dbt_project_evaluator.is_not_empty_string(node.description) | trim ~ " as " ~ dbt.type_boolean() ~ ")",
```

**Review by b-per — CHANGES_REQUESTED** [review]
Thanks for this.
CI is failing but I have added cooment.

File: README.md [hunk 2]
```diff
@@ -22,14 +22,16 @@ Currently, the following adapters are supported:
 - DuckDB
 - Trino (tested with Iceberg connector)
 - AWS Athena (tested manually)
+- Greenplum
```
Comment (graciegoheen) [inline README.md:25]:
```suggestion
- Greenplum (tested manually)
```

File: README.md [hunk 3]
```diff
@@ -22,14 +22,16 @@ Currently, the following adapters are supported:
...
+- Clickhouse
```
Comment (graciegoheen) [inline README.md:26]:
```suggestion
- Clickhouse (tested manually)
```

File: macros/unpack/get_exposure_values.sql (hunk 1 above)
Comment (igorvoltaic) [reply macros/unpack/get_exposure_values.sql]: thanks!

**Comment by igorvoltaic** [comment]
still need to look into broken tests

**Comment by igorvoltaic** [comment]
to run tests locally:

```bash
docker run --rm -d --name ch-server -p8123:8123 --ulimit nofile=262144:262144 clickhouse/clickhouse-server
export CLICKHOUSE_TEST_HOST=localhost
export CLICKHOUSE_TEST_PORT=8123
export CLICKHOUSE_TEST_USER=
export CLICKHOUSE_TEST_PASS=
export CLICKHOUSE_TEST_DBNAME=default

./run_test.sh clickhouse
```

Screenshot 2024-06-13 at 07 59 45
Screenshot 2024-06-13 at 08 00 10
Screenshot 2024-06-13 at 07 59 59

File: models/marts/structure/fct_source_directories.sql [hunk 4]
```diff
@@ -6,7 +6,7 @@
 
 with all_graph_resources as (
     select * from {{ ref('int_all_graph_resources') }}
-    where not is_excluded
+    where not cast(is_excluded as {{ dbt.type_boolean() }})
```
Comment (b-per) [inline models/marts/structure/fct_source_directories.sql:9]: Do we need to cast this one here in Clickhouse?

File: models/marts/core/int_all_graph_resources.sql [hunk 5]
```diff
@@ -23,116 +23,116 @@ with unioned as (
...
 
 calculate_model_type as (
-    select 
-        *, 
-        case 
-            when resource_type in ('test', 'source', 'metric', 'exposure', 'seed') then null
+    select
+        *,
+        case
+            when joined.resource_type in ('test', 'source', 'metric', 'exposure', 'seed') then null
```
Comment (b-per) [inline models/marts/core/int_all_graph_resources.sql:131]: This is not the convention of this project, we don't qualify the name of tables/CTEs when we read just from one. Could you revert back to not qualifying the table/CTE name?

File: seeds/dbt_project_evaluator_exceptions.csv [hunk 6]
```diff
@@ -1 +1,2 @@
-fct_name,column_name,id_to_exclude,comment
\ No newline at end of file
+fct_name,column_name,id_to_exclude,comment
+,,,
```
Comment (b-per) [inline seeds/dbt_project_evaluator_exceptions.csv:2]: Is this change necessary?

**Review by b-per — CHANGES_REQUESTED** [review]
Thanks for this.

We had a look with the team and have a couple of comments/change requests. It should be OK once those are addressed.

Also, we just merged another PR today so there are a couple of merge conflicts to fix.

File: seeds/dbt_project_evaluator_exceptions.csv (hunk 6 above)
Comment (igorvoltaic) [reply seeds/dbt_project_evaluator_exceptions.csv:2]: Yes, absolutely. ClickHouse expects an empty line here and Bigquery fails if there are no commas to separate empty fields

File: models/marts/structure/fct_source_directories.sql (hunk 4 above)
Comment (igorvoltaic) [reply models/marts/structure/fct_source_directories.sql:9]: Actually not, let's remove this

File: models/marts/core/int_all_graph_resources.sql [hunk 7]
```diff
@@ -23,116 +23,116 @@ with unioned as (
...
-        unioned_with_calc.resource_name, 
-        unioned_with_calc.prefix, 
-        unioned_with_calc.resource_type, 
-        unioned_with_calc.file_path, 
-        unioned_with_calc.directory_path,
-        unioned_with_calc.is_generic_test,
-        unioned_with_calc.file_name,
-        case 
+        unioned_with_calc.unique_id as resource_id,
+        unioned_with_calc.resource_name as resource_name,
```
[the remark this answers is not in the document]
Comment (igorvoltaic) [reply models/marts/core/int_all_graph_resources.sql]: removed all unnecessary aliasing

File: macros/recursive_dag.sql [hunk 8]
```diff
@@ -15,7 +15,7 @@ with recursive direct_relationships as (
 
 -- recursive CTE
 -- one record for every resource and each of its downstream children (including itself)
-all_relationships (
+all_relationships {% if target.name not in ['clickhouse'] %} (
```
[the remark this answers is not in the document]
Comment (igorvoltaic) [reply macros/recursive_dag.sql]: moved to separate macro

File: models/marts/core/int_all_graph_resources.sql (hunk 5 above)
Comment (igorvoltaic) [reply models/marts/core/int_all_graph_resources.sql]: removed qualified names where there is no need using them

File: integration_tests/models/marts/fct_model_6.sql [hunk 9]
```diff
@@ -4,7 +4,7 @@
     )
 }}
 
-select 1 as id 
+select {{ 'toInt32(1)' if target.name in ['clickhouse'] else '1' }} as id
```
[the remark this answers is not in the document]
Comment (igorvoltaic) [reply integration_tests/models/marts/fct_model_6.sql]: set to -1234567890 else that was going to be unsigned int

**Comment by igorvoltaic** [comment]
Screenshot 2024-06-23 at 14 08 27
Screenshot 2024-06-23 at 14 08 10
Screenshot 2024-06-23 at 14 07 56

File: seeds/dbt_project_evaluator_exceptions.csv (hunk 6 above)
Comment (igorvoltaic) [reply seeds/dbt_project_evaluator_exceptions.csv]: I've gave it a second thought

File: models/marts/core/int_all_graph_resources.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (igorvoltaic) [reply models/marts/core/int_all_graph_resources.sql]: These are column names you get in older versions (i.e. Clickhouse v23.8).
Screenshot 2024-07-31 at 07 36 08

Most of the downstream models fail because with similar errors because of this
```shell
04:31:28    Database Error in model fct_source_directories (models/marts/structure/fct_source_directories.sql)
  :HTTPDriver for http://localhost:8123 returned response code 404)
   Code: 47. DB::Exception: Missing columns: 'file_name' 'source_name' 'file_path' 'directory_path' while processing query: 'SELECT resource_name, resource_type, file_path AS current_file_path, concat('models/', 'staging', '/', source_name, '/', file_name) AS change_file_path_to FROM all_graph_resources WHERE (resource_type = 'source') AND (directory_path NOT LIKE concat('%', source_name, '%'))', required columns: 'resource_name' 'directory_path' 'resource_type' 'file_path' 'source_name' 'file_name' 'resource_name' 'directory_path' 'resource_type' 'file_path' 'source_name' 'file_name'. (UNKNOWN_IDENTIFIER) (version 23.8.16.40 (official build))
  compiled Code at target/run/dbt_project_evaluator/models/marts/structure/fct_source_directories.sql
```

File: macros/recursive_dag.sql [hunk 10]
```diff
@@ -243,9 +251,9 @@ with direct_relationships as (
         child.directory_path as child_directory_path,
         child.file_name as child_file_name,
         child.is_excluded as child_is_excluded,
-        all_relationships_unioned.distance,
-        all_relationships_unioned.path,
-        all_relationships_unioned.is_dependent_on_chain_of_views
+        cast(all_relationships_unioned.distance as {{ dbt.type_int() }}) as distance,
+        all_relationships_unioned.path as path,
```
Comment (b-per) [inline macros/recursive_dag.sql:255]: As part of our way of working we don't alias name when it doesn't change

File: macros/unpack/get_node_values.sql [hunk 11]
```diff
@@ -23,30 +23,30 @@
...
+                "cast(" ~ node.config.enabled | trim ~ " as " ~ dbt.type_boolean() ~ ")",
                 wrap_string_with_quotes(node.config.materialized),
                 wrap_string_with_quotes(node.config.on_schema_change),
                 wrap_string_with_quotes(node.group),
                 wrap_string_with_quotes(node.access),
                 wrap_string_with_quotes(node.latest_version),
                 wrap_string_with_quotes(node.version),
                 wrap_string_with_quotes(node.deprecation_date),
-                "cast(" ~ contract | trim  ~ " as boolean)",
+                "cast(" ~ contract | trim  ~ " as " ~ dbt.type_boolean() ~ ")",
```
Comment (b-per) [inline macros/unpack/get_node_values.sql:34]: Thanks for updating all those 👍

**Review by b-per — CHANGES_REQUESTED** [review]
A couple of changes would be great to align with the conventions we use. Thanks for the work on this one.

File: integration_tests/models/marts/fct_model_6.sql [hunk 12]
```diff
@@ -4,7 +4,7 @@
     )
 }}
 
-select 1 as id 
+select -1234567890 as id
```
[the remark this answers is not in the document]
Comment (igorvoltaic) [reply integration_tests/models/marts/fct_model_6.sql:7]: I've set relevant type in yml files where it was necessary and `select 1` untouched

File: macros/recursive_dag.sql [hunk 13]
```diff
@@ -15,35 +15,8 @@ with recursive direct_relationships as (
...
-    child_source_name,
-    child_file_path,
-    child_directory_path,
-    child_file_name,
-    child_is_excluded,
-    distance,
-    path,
-    is_dependent_on_chain_of_views
-) as (
+all_relationships {{ recursive_dag__all_relationships_cte_aliases() }}
```
[the remark this answers is not in the document]
Comment (igorvoltaic) [reply macros/recursive_dag.sql:18]: I've removed changes to this macro completely and added a couple aliases to make `bigquery__recursive_dag` work, since that was already required to support clickhouse 24.3 lte

**Review by b-per — APPROVED** [review]
Thanks for the work here!

I will get it released as a new minor in a bit. I'll just squash the commits as there are quite a few of them and many reverting previous changes following the feedback we gave.
